kore/src/Kore/Step/Function/Evaluator.hs

```diff
@@ -79,50 +93,76 @@ evaluateApplication
     -> Application Symbol (TermLike variable)
     -- ^ The pattern to be evaluated
     -> simplifier (OrPattern variable)
-evaluateApplication configurationPredicate childrenPredicate application = do
+evaluateApplication
```

Comment:
It's obviously not this PR's job to fix this, but I'm not a fan of functions as big/complex as this one. Maybe we could create a "refactor" label and make a small issue pointing to individual areas/functions everytime we find such a thing, and we can grab a few of those each sprint?

Reply:
I prefer to refactor when we are already working on the affected code, so I think `TODO`s are the most effective reminder. 

As far as refactoring goes: in this case, I would factor out a memo-less "function evaluator" object in the same style as `Kore.Step.Function.Memo.Self` and add a function

```.hs
memoize :: Evaluator.Self state -> Memo.Self state -> Evaluator.Self state
```

which applies memoization to an evaluator object.